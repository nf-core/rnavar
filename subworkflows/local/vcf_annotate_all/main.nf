//
// ANNOTATION
//

include { VCF_ANNOTATE_BCFTOOLS                         } from '../vcf_annotate_bcftools/main'
include { VCF_ANNOTATE_ENSEMBLVEP                       } from '../../nf-core/vcf_annotate_ensemblvep/main'
include { VCF_ANNOTATE_ENSEMBLVEP as VCF_ANNOTATE_MERGE } from '../../nf-core/vcf_annotate_ensemblvep/main'
include { VCF_ANNOTATE_SNPEFF                           } from '../../nf-core/vcf_annotate_snpeff/main'

workflow VCF_ANNOTATE_ALL {
    take:
    vcf          // channel: [ val(meta), vcf ]
    fasta
    tools        // Mandatory, list of tools to apply
    snpeff_db
    snpeff_cache
    vep_genome
    vep_species
    vep_cache_version
    vep_cache
    vep_extra_files

    main:
    def ch_reports = Channel.empty()
    def vcf_annotated = Channel.empty()
    def tab_annotated = Channel.empty()
    def json_annotated = Channel.empty()
    def ch_versions = Channel.empty()

    if (tools.split(',').contains('bcfann')) {
        VCF_ANNOTATE_BCFTOOLS(vcf)

        vcf_annotated = vcf_annotated.mix(VCF_ANNOTATE_BCFTOOLS.out.vcf_tbi)
        ch_versions = ch_versions.mix(VCF_ANNOTATE_BCFTOOLS.out.versions)
    }


    if (tools.split(',').contains('merge') || tools.split(',').contains('snpeff')) {
        VCF_ANNOTATE_SNPEFF(vcf, snpeff_db, snpeff_cache)

        ch_reports = ch_reports.mix(VCF_ANNOTATE_SNPEFF.out.reports.map{ _meta, reports_ -> [ reports_ ] })
        vcf_annotated = vcf_annotated.mix(VCF_ANNOTATE_SNPEFF.out.vcf_tbi)
        ch_versions = ch_versions.mix(VCF_ANNOTATE_SNPEFF.out.versions)
    }

    if (tools.split(',').contains('merge')) {
        vcf_ann_for_merge = VCF_ANNOTATE_SNPEFF.out.vcf_tbi.map{ meta, vcf_, _tbi -> [ meta, vcf_, [] ] }
        VCF_ANNOTATE_MERGE(vcf_ann_for_merge, fasta, vep_genome, vep_species, vep_cache_version, vep_cache, vep_extra_files)

        ch_reports = ch_reports.mix(VCF_ANNOTATE_MERGE.out.reports)
        vcf_annotated = vcf_annotated.mix(VCF_ANNOTATE_MERGE.out.vcf_tbi)
        ch_versions = ch_versions.mix(VCF_ANNOTATE_MERGE.out.versions)
    }

    if (tools.split(',').contains('vep')) {
        vcf_for_vep = vcf.map{ meta, vcf_ -> [ meta, vcf_, [] ] }
        VCF_ANNOTATE_ENSEMBLVEP(vcf_for_vep, fasta, vep_genome, vep_species, vep_cache_version, vep_cache, vep_extra_files)

        ch_reports  = ch_reports.mix(VCF_ANNOTATE_ENSEMBLVEP.out.reports)
        vcf_annotated  = vcf_annotated.mix(VCF_ANNOTATE_ENSEMBLVEP.out.vcf_tbi)
        tab_annotated  = tab_annotated.mix(VCF_ANNOTATE_ENSEMBLVEP.out.tab)
        json_annotated = json_annotated.mix(VCF_ANNOTATE_ENSEMBLVEP.out.json)
        ch_versions = ch_versions.mix(VCF_ANNOTATE_ENSEMBLVEP.out.versions)
    }

    emit:
    vcf_ann = vcf_annotated     // channel: [ val(meta), vcf.gz, vcf.gz.tbi ]
    tab_ann = tab_annotated
    json_ann = json_annotated
    reports = ch_reports      //    path: *.html
    versions = ch_versions     //    path: ch_versions.yml
}
