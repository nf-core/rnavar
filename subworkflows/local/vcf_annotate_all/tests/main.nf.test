nextflow_workflow {

    name "Test Subworkflow VCF_ANNOTATE_ALL"
    script "../main.nf"
    workflow "VCF_ANNOTATE_ALL"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/vcf_annotate_all"
    tag "ensemblvep/vep"
    tag "snpeff/snpeff"
    tag "subworkflows/vcf_annotate_ensemblvep"
    tag "subworkflows/vcf_annotate_snpeff"

    options "-stub"

    setup {
        run("ENSEMBLVEP_DOWNLOAD") {
            script "../../../../modules/nf-core/ensemblvep/download/main.nf"

            process {
                """
                input[0] = Channel.of([
                    [id:"cache"],
                    params.vep_genome,
                    params.vep_species,
                    params.vep_cache_version
                ])
                """
            }
        }

        run("SNPEFF_DOWNLOAD") {
            script "../../../../modules/nf-core/snpeff/download/main.nf"

            process {
                """
                input[0] = Channel.of([
                    [id:"cache"],
                    params.snpeff_db
                ])
                """
            }
        }
    }

    test("vep") {
        when {
            params {
                annotate_tools = "vep"
                vep_genome = "WBcel235"
                vep_species = "caenorhabditis_elegans"
                vep_cache_version = "110"
                snpeff_db = "WBcel235.105"
            }

            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf', checkIfExists: true)
                ])
                input[1] = Channel.value([
                    [id:"fasta"],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                input[2] = params.annotate_tools
                input[3] = params.snpeff_db             // snpeff_db
                input[4] = []                           // snpeff_cache
                input[5] = params.vep_genome            // vep_genome
                input[6] = params.vep_species           // vep_species
                input[7] = params.vep_cache_version     // vep_cache_version
                input[8] = ENSEMBLVEP_DOWNLOAD.out.cache.map { meta, cache -> cache } // vep_cache
                input[9] = Channel.value([])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("snpeff") {
        when {
            params {
                annotate_tools = "snpeff"
                vep_genome = "WBcel235"
                vep_species = "caenorhabditis_elegans"
                vep_cache_version = "110"
                snpeff_db = "WBcel235.105"
            }

            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf', checkIfExists: true)
                ])
                input[1] = Channel.value([
                    [id:"fasta"],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                input[2] = params.annotate_tools
                input[3] = params.snpeff_db             // snpeff_db
                input[4] = SNPEFF_DOWNLOAD.out.cache    // snpeff_cache
                input[5] = params.vep_genome            // vep_genome
                input[6] = params.vep_species           // vep_species
                input[7] = params.vep_cache_version     // vep_cache_version
                input[8] = []                           // vep_cache
                input[9] = Channel.value([])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("merge") {
        when {
            params {
                annotate_tools = "merge"
                vep_genome = "WBcel235"
                vep_species = "caenorhabditis_elegans"
                vep_cache_version = "110"
                snpeff_db = "WBcel235.105"
            }

            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf', checkIfExists: true)
                ])
                input[1] = Channel.value([
                    [id:"fasta"],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                input[2] = params.annotate_tools
                input[3] = params.snpeff_db             // snpeff_db
                input[4] = SNPEFF_DOWNLOAD.out.cache    // snpeff_cache
                input[5] = params.vep_genome            // vep_genome
                input[6] = params.vep_species           // vep_species
                input[7] = params.vep_cache_version     // vep_cache_version
                input[8] = ENSEMBLVEP_DOWNLOAD.out.cache.map { meta, cache -> cache } // vep_cache
                input[9] = Channel.value([])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
