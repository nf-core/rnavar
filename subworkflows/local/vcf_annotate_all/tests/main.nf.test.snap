{
    "vep": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_VEP.ann.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                        "test_VEP.ann.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    "test_VEP.ann_summary.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "4": [
                    "versions.yml:md5,6b4e45ba6bc1085f28e972f10c018f77",
                    "versions.yml:md5,6c3f1b0776837530cda145392ab8f18b"
                ],
                "json_ann": [
                    
                ],
                "reports": [
                    "test_VEP.ann_summary.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "tab_ann": [
                    
                ],
                "vcf_ann": [
                    [
                        {
                            "id": "test"
                        },
                        "test_VEP.ann.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                        "test_VEP.ann.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,6b4e45ba6bc1085f28e972f10c018f77",
                    "versions.yml:md5,6c3f1b0776837530cda145392ab8f18b"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-04T16:19:31.736660565"
    },
    "merge": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "test_VEP.ann.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                        "test_VEP.ann.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    [
                        "test_snpEff.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    "test_VEP.ann_summary.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "4": [
                    "versions.yml:md5,4a322b8b0f7017858530180a68232e96",
                    "versions.yml:md5,6fb558faa1c4c79e757970991dfbf118",
                    "versions.yml:md5,abf2b743072f2ba8b250b3925061622f",
                    "versions.yml:md5,d5804f2a2b5907e982d54dca0ba2db44"
                ],
                "json_ann": [
                    
                ],
                "reports": [
                    [
                        "test_snpEff.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    "test_VEP.ann_summary.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "tab_ann": [
                    
                ],
                "vcf_ann": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "test_VEP.ann.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                        "test_VEP.ann.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,4a322b8b0f7017858530180a68232e96",
                    "versions.yml:md5,6fb558faa1c4c79e757970991dfbf118",
                    "versions.yml:md5,abf2b743072f2ba8b250b3925061622f",
                    "versions.yml:md5,d5804f2a2b5907e982d54dca0ba2db44"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-04T16:32:38.738793297"
    },
    "snpeff": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    [
                        "test_snpEff.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "versions.yml:md5,4a322b8b0f7017858530180a68232e96",
                    "versions.yml:md5,d5804f2a2b5907e982d54dca0ba2db44"
                ],
                "json_ann": [
                    
                ],
                "reports": [
                    [
                        "test_snpEff.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "tab_ann": [
                    
                ],
                "vcf_ann": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,4a322b8b0f7017858530180a68232e96",
                    "versions.yml:md5,d5804f2a2b5907e982d54dca0ba2db44"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-04T16:32:26.089637442"
    }
}